### PR TITLE
[Feature] Add run info

### PR DESCRIPTION
**Pull request by Corrade**
image
(This screenshot shows the latest UI)

## What are the changes?
Per-run stats!
- Knockouts, assists, faints, damage dealt, damage taken (blocked)
- The menu can be pulled up mid-run and is also shown after a run ends
- Persistence via session data

See the "[#2374] Run info" thread in the Discord's #dev-corner channel for more discussion.

## Why am I doing these changes?
- Provides a much-needed recap at the end of a run
- Low-risk, innocuous feature that's easily ignored

## What did change?
- Added runData class to Pokemon and PokemonData classes
- Modified Pokemon::damage() and its callers to maintain runData. This was a top-down approach that benefited from the fact that Pokemon::damage() is consistently used for all damage-related actions.
- Added sourceId field to Weather and Status to support damage attribution
- Added sourceId field to ConfusedTag, NightmareTag and PerishSongTag to support damage attribution
- Added runInfoUiHandler to the menuUiHandler
- Hooked up runInfoUiHandler fields to localisation
- Added EndRunInfoPhase to show the run info UI at the end of a run

### What's changed since the GIFs below were recorded
- Renamed from 'run stats' to 'run info' in anticipation of future non-stat extensions
- Elevated menu option from below 'stats' to below 'game settings' as I presume this UI would be more frequently opened
- Renamed stats, e.g. from kills (K) to knockouts (K.O.), as shown in the topmost screenshot here
- Flipped rows and columns in the UI to support longer stat labels
- Removed Pokemon names from the UI

### Screenshots/Videos
- Basic demo
run_stats_basic
- Run info UI appears at the end of the game
run_stats_end

#### Damage attribution
- Works with OHKO moves
run_stats_ohko
- Works with weather (credits the weather setter)
run_stats_weather
- Works with non-volatile status effects
run_stats_non_volatile_status
- Works with abilities (e.g. rough skin)
run_stats_abilities
- Works with arena tags (e.g. stealth rock)
run_stats_arena_tags
- Works with battler tags (e.g. confusion)
run_stats_battler_tags
- Works with pain split
run_stats_pain_split

#### Assists
- A Pokemon gets an assist if it was present in the fight or dealt damage (e.g. via stealth rock)
- Exception: at the start of an encounter, a lead that gets immediately swapped out does not get assist credit
run_stats_immediate_swaps_are_uncredited
- Exception: in a double battle, a Pokemon switching in does not get credit for an opponent who dies on the same turn of the switch
run_stats_switchins_are_uncredited_for_same_turn_kills
- Assists are granted even to Pokemon that faint throughout the battle
run_stats_assists_can_be_credited_to_fainted_pokemon

#### Policies
- Healing is ignored for now
- Self-kills, assists and damage-dealt are ignored, though self-damage-taken and deaths are still counted
run_stats_self_damage
- Friendly fire counts; damage and kills from friendly pokemon (i.e. in doubles) count
run_stats_friendly_fire
- Evolution doesn't reset run info
run_stats_persistence_through_evolution
- Form changes don't reset run info
run_stats_persistence_through_form_change
- Fusing/splicing combines run info
run_stats_fusion
- Newly caught Pokemon start with zero run info (run info isn't tracked for opponents)
run_stats_zero_for_new_catches

#### Implementation
- Run info is tied to sessions (runs)
run_stats_persistence
- The new sourceId field in the Weather class persists (Porygon is still credited with the hail damage and the assist after a reload)
run_stats_weather_source_persistence
- The new sourceId field in the Status class persists (Porygon is still credited with the toxic damage after a reload; Porygon dealt a total of 4 damage = 1 from the first tick + 3 from the second post-reload tick)
run_stats_status_source_persistence
- The run info UI can't be seen from the main menu i.e. outside a run
run_stats_invisible_on_main_menu

## How to test the changes?
- The core feature can be tested by playing the game regularly and looking at the run info UI
- The above edge cases can be replicated using appropriate overrides
- No automated tests (WIP)

## Checklist
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes (manually)?
    - [x] Are all unit tests still passing? (`npm run test`)
- [x] Are the changes visual?
  - [x] Have I provided screenshots/videos of the changes?

### CODE REVIEW

File: src/locales/en/run-stats-ui-handler.ts
Comment (Corrade) [inline src/locales/en/run-stats-ui-handler.ts:1]: Walker: Kills -> knockouts or takedowns? KOs?
Cat Capitalist: Kills -> wins?
Nobody is a big fan of MIT.

File: src/field/pokemon.ts [hunk 1]
```diff
@@ -2051,18 +2057,62 @@ export default abstract class Pokemon extends Phaser.GameObjects.Container {
...
+        this.scene.unshiftPhase(new FaintPhase(this.scene, this.getBattlerIndex(), preventEndure));
+        this.resetSummonData();
+      }
     }
 
     return damage;
   }
 
-  damageAndUpdate(damage: integer, result?: DamageResult, critical: boolean = false, ignoreSegments: boolean = false, preventEndure: boolean = false, ignoreFaintPhase: boolean = false): integer {
+  damageAndUpdate(damage: integer, result?: DamageResult, sourcePokemon?: Pokemon, critical: boolean = false, ignoreSegments: boolean = false, preventEndure: boolean = false, ignoreFaintPhase: boolean = false): integer {
```
Comment (Corrade) [inline src/field/pokemon.ts:2112]: I'm tempted to make this new sourcePokemon parameter the second argument and non-optional. Every caller currently has a sourcePokemon to provide, whereas not every caller provides a result parameter.

File: src/locales/en/run-stats-ui-handler.ts
Comment (Corrade) [reply src/locales/en/run-stats-ui-handler.ts:1]: Done!

**Comment by brain-frog** [comment]
I think this is a fun idea but I don't want to put anything else into the menu right now since it is not yet scroll-able (I know that can be added soon but there's other priorities). 
Personally I think this should be part of the run history feature. I don't think people need to see their stats mid run, but it would be good to show these stats after a run ends. I'm willing to discuss this further if you'd like.

**Comment by Corrade** [comment]
@brain-frog Thanks a lot for the note!

I've been talking to the run history contributor (cat capitalist) and he's already onboard with including this display in his UI!

Obviously I'm biased and think it'd be fun to check these stats mid-run. I understand that on one hand, it doesn't really inform any decision-making. On the other, people enjoy looking at stats in other games, namely multiplayer ones, and whilst we're obviously in a different genre I could see people having the same interest here.

I'm happy to make the main menu scrollable if that's a blocker. It has precedent in the settings menu so I don't think it'll be a big lift.

Though, if you'd still like to keep this UI within the run history, no worries - I can fold this PR into cat capitalist's run history too.

**Comment by Tempo-anon** [comment]
There's a lot of conflicts here now and run history has been implemented. Is there any interest in continuing to build this in there?

**Comment by Corrade** [comment]
> There's a lot of conflicts here now and run history has been implemented. Is there any interest in continuing to build this in there?

Looking into it!

**Comment by flx-sta** [comment]
@Corrade I'm putting this into Draft while you are working on the conflicts.

I'm also wondering if this might actually be obsolete after adding the run-history
